Make sidebar slide in and slide out

dojutsu-user opened the pull request:
Fixes #697 

![dimpledmediocreherald-small](https://user-images.githubusercontent.com/29149191/47966844-db9f0b00-e07c-11e8-931c-9c2b63ec5abf.gif)

dojutsu-user commented:
I am still unsure about these lines

https://github.com/rtfd/sphinx_rtd_theme/blob/4a3d790e61d7524baac698badee814cd286b1219/sass/_theme_layout.sass#L372-L375

Blendify commented:
https://stackoverflow.com/questions/9271276/is-the-recommendation-to-include-css-before-javascript-invalid

Says it does not matter except for mobile where the difference is very small and speculative parsing will probably come to mobile at some point anyway.

Blendify commented:
I think this makes sense but one additional thing that should be added is swipe gesture support

Seems pretty simple, here is a jquery example: http://demos.jquerymobile.com/1.3.2/examples/panels/panel-swipe-open.html#&ui-state=dialog

jessetan commented on sass/_theme_layout.sass:
```diff
@@ -370,12 +371,15 @@ footer
-      left: -$nav-desktop-width
+      transform: $sidebar-transform-hide
+      transition: $sidebar-transition
     @else
```
The `$nav-desktop-position` is already broken without sliding, no need to worry about this line I think

jessetan commented on js/theme.js:
```diff
@@ -238,3 +238,33 @@ if (typeof(window) != 'undefined') {
+          var offset = 100;
+          if(start) {
+              var end = event.changedTouches.item(0).clientX;
+              if(end > start + offset) {
```
I'm personally not in favor of expanding the menu using swipes. It is too easy to trigger when using the page (e.g. selecting text). Closing it by swiping is OK

jessetan commented on js/theme.js:
```diff
@@ -238,3 +238,33 @@ if (typeof(window) != 'undefined') {
+// swiping right/close.
+(function () {
+    var start = null;
+      window.addEventListener("touchstart",function(event){
```
These functions should go into `ThemeNav` and the listeners added during the `init()` call

jessetan commented on sass/_theme_variables.sass:
```diff
@@ -52,6 +52,11 @@ $nav-link-color-visited:              $purple
+// Sidebar
+$sidebar-transform-hide:              translateX(-100%)
+$sidebar-transform-show:              translateX(0%)      
+$sidebar-transition:                  transform 0.3s ease
```
Transforms and transitions may need prefixed forms as well for IE 9 and older Android and iOS browsers

dojutsu-user commented:
@jessetan 
I think I have added all the improvements suggested by you.
Also, I have changed the javascript code to jquery, implementing the `swiperight` and `swipeleft` events, but for some reason **it is not working**.

dojutsu-user commented:
@jessetan 
Added the improvements.
I googled a little bit and found that maximum allowed time for the swiping so that the sidebar slides in can be added.
I have added it.
Need you review.

dojutsu-user commented on sass/_theme_badge.sass:
```diff
@@ -85,6 +85,8 @@
+    @include prefixer(transition, $sidebar-transition, webkit moz ms o spec)
     &.shift
       display: block
+      transform:  @include prefixer(transform, $sidebar-transform-show, webkit moz ms o spec)
```
I totally forgot about this.
Thanks

jessetan commented on js/theme.js:
```diff
@@ -97,6 +105,31 @@ function ThemeNav () {
+                }
+            })
+            .on('touchend', ".wy-body-for-nav", function(event) {
+                if (self.sidebarSwipe.startX) {
```
```suggestion
                if (self.sidebarSwipe.startX === null) {
```
because startX can have a value of `0`

dojutsu-user replied:
Done.
Thanks

dojutsu-user replied:
I just tested it locally, and it is not working if the suggesstion is applied.

jessetan replied:
Whoops, should be `!== null`

jessetan commented:
If I run this in desktop Chrome devtools with device emulation, my vertical scrolling is broken when the menu is visible. Does it work for you (on a real device)?

jessetan commented:
> Exactly what you mean by this?

I mean that I can swipe up and down to scroll the body content when the menu is collapsed, but when the menu is open I can not swipe up and down to scroll the body content nor the contents of the menu if it is longer than can fit on the screen.

dojutsu-user commented:
@jessetan 
When the menu is open, swiping up/down on the body is not there. But I can swipe up/down on the menu if it is longer than the screen size (tested on a real device).

![ezgif-4-68eed51cc744](https://user-images.githubusercontent.com/29149191/50337647-cea87f00-0537-11e9-9361-d85ab5bcda16.gif)

dojutsu-user commented:
@jessetan 
Any update on this?

jessetan commented:
Expanding the menu on the "Paragraph level markup" page in the the documentation (https://sphinx-rtd-theme.readthedocs.io/en/latest/demo/demo.html) makes the content part invisible. This works if I remove the `position: fixed` on `.wy-nav-content-wrap.shift` but there may be other side effects.

Is it possible to close the menu when the content part is clicked/tapped? That feels more logicol on mobile

With this PR I can't seem to do inertial scrolling (is that the name?) on mobile Chrome or Safari where you give it a fast flick and it keeps scrolling for a while
